packages/react/src/components/Modal/ModalProvider.tsx

```diff
@@ -0,0 +1,159 @@
+    props?: Partial<DeleteConfirmationModalProps>,
+  ) => Promise<Decision>;
+  confirmationModal: (
+    props?: Partial<ConfirmationModalProps>,
+  ) => Promise<Decision>;
+  messageModal: (props?: Partial<MessageModalProps>) => Promise<Decision>;
+}
+
+const ModalContext = createContext<undefined | ModalContextType>(undefined);
+
+export const useModals = () => {
+  const context = useContext(ModalContext);
+  if (context === undefined) {
+    throw new Error("useModals must be used within a ModalProvider.");
+  }
+  return context;
+};
+
+/**
+ * This component is used to wrap a modal component and add the modal context to it.
+ * It handles creating a generic `component` modal, and closing it when the `onClose` callback is called.
+ *
+ * @param component
+ * @param onClose
+ * @param props
+ * @constructor
+ */
+const ModalContainer = ({
+  component,
+  onClose,
+  props,
+}: {
+  component: DecisionModalComponent;
+  onClose: Function;
+  props: Partial<DecisionModalProps>;
+}) => {
+  const modal = useModal({ isOpenDefault: true });
+
+  const close = () => {
+    modal.onClose();
+    onClose();
+  };
+
+  const onCloseLocal = () => {
+    close();
+    // Closing sends a "undefined" decision.
+    props.onDecide?.();
+  };
+
+  const onDecide = (decision?: Decision) => {
+    close();
+    props.onDecide?.(decision);
+  };
+
+  return createElement(component, {
+    ...modal,
+    ...props,
+    onClose: onCloseLocal,
+    onDecide,
+  });
+};
+
+type ModalMap = Map<string, ReactNode>;
+
+export const ModalProvider = ({ children }: PropsWithChildren) => {
+  const [modals, setModals] = useState<ModalMap>({} as ModalMap);
+
+  const addModal = (
+    component: FunctionComponent<DecisionModalProps>,
+    props: Partial<DecisionModalProps>,
+  ) => {
+    const key = randomString(32);
+    const container = (
+      <ModalContainer
+        component={component}
+        props={props}
+        onClose={() => {
+          setModals((modals_) => {
+            // @ts-ignore
+            delete modals_[key];
+            return modals_;
+          });
+        }}
+      />
+    );
+    setModals((modals_) => ({
+      ...modals_,
+      [key]: container,
+    }));
+  };
+
+  const ModalHelper = (component: DecisionModalComponent) => {
+    return (props: Partial<DecisionModalProps> = {}) => {
+      return new Promise<Decision>((resolve) => {
+        addModal(component, {
+          onDecide: (decision) => {
+            resolve(decision);
+          },
+          ...props,
+        });
+      });
+    };
+  };
+
+  const context: ModalContextType = useMemo(
+    () => ({
+      deleteConfirmationModal: ModalHelper(DeleteConfirmationModal),
+      confirmationModal: ModalHelper(ConfirmationModal),
+      messageModal: ModalHelper(MessageModal),
+    }),
+    [],
+  );
+
+  return (
+    <ModalContext.Provider value={context}>
+      {children}
+      <div id="c__modals-portal" />
+      {Object.entries(modals).map(([key, modal]) => (
```

Comment:
I wonder if you could not use `React.Children.map)` instead.
https://react.dev/reference/react/Children#children-map

Reply:
Good idea, I tried and `Children` is of type `ReactNode`and `modals` is of type `Map` so it does not become really handy and this syntax does not work: 


```
{React.Children.map(modals, ([key, modal]) => (
        <Fragment key={key}>{modal}</Fragment>
      ))}

```

😅